### PR TITLE
Enhance `check_version`

### PR DESCRIPTION
**Pull request by layday**

### CODE REVIEW

File: src/build/__init__.py
Comment (gaborbernat) [inline src/build/__init__.py:75]: Could we perhaps use somehow ``pip check`` instead?

Comment (layday) [reply src/build/__init__.py:75]: Well, that would place a runtime dependency on pip (we could use `ensurepip` I suppose but distros might then be inclined to patch `build` to 'unvendor' pip), and unless we want to be calling its internal API we're looking at a ~400ms overhead per check purely in import time.  IUC pip does not keep track of extras so `pip check` would not error for e.g. `setuptools_scm[toml]` if the toml extra dependencies weren't available.

Comment (gaborbernat) [reply src/build/__init__.py:75]: We already have pip as a runtime dependency - https://github.com/pypa/build/blob/master/src/build/env.py#L173-L188; and this would allow us to drop packaging as install dependency.

Comment (layday) [reply src/build/__init__.py:75]: Even then, we'd have to combine `check` with `list` or `show` to verify that the build dependencies are installed and `list` is the only one of those commands which produces machine-readable data.  I don't know how we'd verify that extras are satisfied with pip.

Comment (FFY00) [reply src/build/__init__.py:70]: The current runtime dependency on pip is for the environments, and it's not even required if ensurepip works. This adds a dependency on pip for `--no-isolation` which we most definitely do not want, that was the whole point of this routine.

**Comment by layday** [comment]
I've made a few changes to return the dependency chain similar to how `pip check` does it.

File: src/build/__init__.py
Comment (FFY00) [inline src/build/__init__.py:61]: Please keep the original import, instead of using a partial import.

File: src/build/__main__.py
Comment (FFY00) [inline src/build/__main__.py:58]: Why change the message string here ("missing" -> "unmet")?

I am not really a fan of the way this is displayed in the CLI:
```
ERROR Unmet dependencies:
	apispec[validation] -> prance[osv] (>=0.11) ; extra == 'validation' -> openapi-spec-validator (>0.2,>=0.2.1) ; extra == 'osv'
```

I think the following would be better:
```
ERROR Unmet dependencies:
	apispec[validation]
        prance[osv] (>=0.11) -> openapi-spec-validator (>0.2,>=0.2.1)
```

File: src/build/__init__.py
Comment (FFY00) [inline src/build/__init__.py:52]: I think it would be better to put effort in upstreaming this into pypa/packaging first, as proposed in pypa/packaging#317.

File: tests/test_projectbuilder.py
Comment (FFY00) [inline tests/test_projectbuilder.py:86]: If we are gonna merge this here, we need to also test recursive extra dependencies.

File: tests/test_projectbuilder.py
Comment (FFY00) [inline tests/test_projectbuilder.py:174]: Why? This is unrelated and just dirties git blame.

**Review by FFY00 — CHANGES_REQUESTED** [review]

File: tests/test_projectbuilder.py
Comment (layday) [reply tests/test_projectbuilder.py:174]: Because `importlib_metadata.Distribution` is mocked further up and `something` is would return a `Distribution` instance, i.e. the dependency would be satisfied.

File: src/build/__init__.py
Comment (layday) [reply src/build/__init__.py:52]: The impression I got was that they were reluctant to add anything which manipulates distributions.  `distlib` would probably be a good fit but everybody seems to be moving away from it.

File: src/build/__main__.py
Comment (layday) [reply src/build/__main__.py:58]: > Why change the message string here ("missing" -> "unmet")?

'Missing' might lead one to think that the dependency can't be located when its version is incompatible.

> I think the following would be better:

It's not clear that the unmet dependency here is `openapi-spec-validator`.  Maybe something like this would be better?

```
ERROR Unmet dependencies:
	openapi-spec-validator (>0.2,>=0.2.1) ; extra == 'osv'
		from prance[osv] (>=0.11) ; extra == 'validation'
				from apispec[validation]
```

File: src/build/__init__.py
Comment (gaborbernat) [reply src/build/__init__.py:61]: We talked about this earlier and for what it's worth, I'm still strongly -1 on original imports and +1 on partial imports. However 🤷

File: src/build/__init__.py
Comment (gaborbernat) [reply src/build/__init__.py:52]: In my experience there's a long cycle with adding stuff to distlib... so if we want it soon we better add it here. We can always remove it if we do manage to upstream it.

File: src/build/__init__.py
Comment (gaborbernat) [reply src/build/__init__.py:70]: Seemed a worthwhile topic to bring up. Considering pip already has a checking functionality, perhaps we don't want to reinvent the wheel.

File: src/build/__init__.py
Comment (layday) [reply src/build/__init__.py]: Restored.

File: tests/test_projectbuilder.py
Comment (layday) [reply tests/test_projectbuilder.py:102]: Added.

File: src/build/__main__.py
Comment (FFY00) [reply src/build/__main__.py:58]: > 'Missing' might lead one to think that the dependency can't be located when its version is incompatible.

The requirement string would include the version requirement, so I find that very unlikely. I do not think this is reason enough to change.

> It's not clear that the unmet dependency here is `openapi-spec-validator`. Maybe something like this would be better?

Well, it is certainly not optimal but I think it's clear enough. Your proposed way does not work great from long chains, I see two issues with it: 1) it puts all dependencies on a new line, which works very badly on >5 chains, and 2) it puts the actual requirement in the end, which is hard to find. It also has the `; extra = ...` which is completely avoidable.

As a user, what I am most interested in knowing is the 1st level missing dependency, in this case `apispec[validation]`, so that I can `pip install apispec[validation]`. As a distro packager, I want to know the actual dependency in the tree, and that is clear enough to me in the format I proposed. We should optimize for the general user.

So, for the interface here I have the following requirements:
  - make it clear, and easily accessible, what is 1st level missing dependency (in this case `apispec[validation]`)
  - provide information about the missing dependency chain (but do not optimize for this)
  - make sure the output does not become unreadable on long chains

File: src/build/__init__.py
Comment (FFY00) [reply src/build/__init__.py:52]: I am in no hurry to merge this as it only affects people using `--no-isolation` when trying to verify recursive extras (extras that depend on extras), which is mostly distro packagers, for whom this is only a nice to have. I would rather put the effort into merging somewhere this could live long term rather than here. Before merging it here, we should figure out what would it take to put it somewhere else and then make the call.

Comment (gaborbernat) [reply src/build/__init__.py:52]: > I am in no hurry to merge this as it only affects people using `--no-isolation`

As a (new) contributor to this project, this sentence on its own would be enough for me to abandon spending any more effort on this PR.

File: src/build/__main__.py
Comment (layday) [reply src/build/__main__.py]: I've changed the formatting as suggested.

File: tests/test_projectbuilder.py
Comment (layday) [reply tests/test_projectbuilder.py]: I've removed the assignment to `importlib_metadata` from the global scope and reverted this change.

File: src/build/__init__.py
Comment (layday) [reply src/build/__init__.py:52]: > as it only affects people ... trying to verify recursive extras

`check_version` would not verify transitive dependencies at all, whether or not they came from extras.

Comment (FFY00) [reply src/build/__init__.py:52]: Yes sorry, my bad.

**Review by FFY00 — APPROVED** [review]
LGTM, let's wait a few days to see if we can hear something from pypa/packaging#317.

**Comment by layday** [comment]
While we are waiting, would anybody know why the pypy3 tests are failing?

**Comment by gaborbernat** [comment]
> While we are waiting, would anybody know why the pypy3 tests are failing?

No, cold you investigate?

**Comment by layday** [comment]
It looks like it's flaky - it passed on Ubuntu this time.  Appears to be coverage-related, possibly https://github.com/nedbat/coveragepy/issues/1010, judging from this line:

```
E                   _sqlite3.InterfaceError: Error binding parameter 0 - probably unsupported type.
```

I don't know why it'd appear now, there might be some interaction with the `monkeypatch` fixture which I replaced the global overwrite of `importlib_metadata.Distribution` with?

**Comment by layday** [comment]
I dropped the commit which enabled branch coverage and that seems to have fixed it (or perhaps not).  We can add branch cov in a separate commit.

**Comment by layday** [comment]
Should we merge this now?

**Review by gaborbernat — APPROVED** [review]

**Comment by FFY00** [comment]
I specifically asked to wait, please do merge things like this without getting the ack from everybody that raised concerns.

**Comment by layday** [comment]
Well, I merely asked if we should merge it - that is, if it would've been time to merge it.  It wasn't an invitation to merge it - sorry if I'd not made myself understood.

**Comment by FFY00** [comment]
The message was directed to Bernát, not you.

**Comment by gaborbernat** [comment]
Sorry, thought we moved past that.

**Pull request merged**
